testplan/runnable/base.py

```diff
@@ -393,106 +394,130 @@ def schedule(self, task=None, resource=None, uid=None, **options):
         :param task: :py:class:`~testplan.runners.pools.tasks.base.Task`
         :param resource: Target pool resource.
         :param resource: :py:class:`~testplan.runners.pools.base.Pool`
-        :param uid: Optional uid for task.
-        :param uid: ``str``
         :param options: Task input options.
         :param options: ``dict``
         :return uid: Assigned uid for task.
         :rtype: ``str``
         """
-        return self.add(
-            task or Task(uid=uid, **options), resource=resource, uid=uid
-        )
+        return self.add(task or Task(**options), resource=resource)
 
-    def add(self, runnable, resource=None, uid=None):
+    def add(self, runnable, resource=None):
         """
-        Adds a
-        :py:class:`runnable <testplan.common.entity.base.Runnable>`
-        tests entity to a :py:class:`~testplan.runners.base.Executor`
-        resource.
+        Adds a :py:class:`runnable <testplan.common.entity.base.Runnable>`
+        tests entity to a :py:class:`~testplan.runners.base.Executor` resource.
 
         :param runnable: Test runner entity.
         :type runnable: :py:class:`~testplan.common.entity.base.Runnable`
         :param resource: Test executor resource.
         :type resource: :py:class:`~testplan.runners.base.Executor`
-        :param uid: Optional test uid.
-        :type uid: ``str``
         :return: Assigned uid for test.
-        :rtype: ``str``
+        :rtype: ``str`` or ```NoneType``
         """
-        uid = uid or getattr(runnable, "uid", strings.uuid4)()
-        if uid in self._tests:
-            self.logger.error(
-                "Skip adding {} with uid {}.. already added.".format(
-                    runnable, uid
-                )
+        local_runner = self.resources.first()
+        resource = resource or local_runner
+
+        if resource not in self.resources:
+            raise RuntimeError(
+                'Resource "{}" does not exist.'.format(resource)
             )
-            return uid
 
-        if isinstance(runnable, Entity):
-            runnable.cfg.parent = self.cfg
+        # The runnable added into TestRunner can be: 1> a real test entity
+        # 2> a task wraps a test entity 3> a callble returns a test entity
+        if isinstance(runnable, Runnable):
```

Comment:
I see that this part of code is dup with what we are doing in _verify_test_target. Can we consolidate them?

Reply:
yes, the code is moved into _verify_test_target(), to make `add` method more short.